c/src/metta.rs

```diff
@@ -569,3 +939,169 @@ pub extern "C" fn metta_load_module(metta: *mut metta_t, name: *const c_char) {
+// =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
+
+/// @brief Renders the config_dir path from the platform environment into a text buffer
+/// @ingroup environment_group
+/// @param[out]  buf  A buffer into which the text will be written
+/// @param[in]  buf_len  The maximum allocated size of `buf`
+/// @return The length of the path string, minus the string terminator character.  If
+/// `return_value > buf_len + 1`, then the text was not fully written and this function should be
+/// called again with a larger buffer.  This function will return 0 if there is no config_dir.
+///
+#[no_mangle]
+pub extern "C" fn environment_config_dir(buf: *mut c_char, buf_len: usize) -> usize {
+    match Environment::platform_env().config_dir() {
+        Some(path) => write_into_buf(path.display(), buf, buf_len),
+        None => 0
+    }
+}
+
+/// @brief Returns the number of module search paths in the environment
+/// @ingroup environment_group
+/// @return The number of paths in the Environment
+///
+#[no_mangle]
+pub extern "C" fn environment_search_path_cnt() -> usize {
+    Environment::platform_env().modules_search_paths().count()
+}
+
+/// @brief Renders nth module search path from the platform environment into a text buffer
+/// @ingroup environment_group
+/// @param[in]  idx  The index of the search path to render, with 0 being the highest priority search
+///     path, and subsequent indices having decreasing search priority.
+///     If `i > environment_search_path_cnt()`, this function will return 0.
+/// @param[out]  buf  A buffer into which the text will be written
+/// @param[in]  buf_len  The maximum allocated size of `buf`
+/// @return The length of the path string, minus the string terminator character.  If
+///     `return_value > buf_len + 1`, then the text was not fully written and this function should be
+///     called again with a larger buffer.  This function will return 0 if the input arguments don't
+///     specify a valid search path.
+///
+#[no_mangle]
+pub extern "C" fn environment_nth_search_path(idx: usize, buf: *mut c_char, buf_len: usize) -> usize {
+    let path = Environment::platform_env().modules_search_paths().nth(idx);
+    match path {
+        Some(path) => write_into_buf(path.display(), buf, buf_len),
+        None => 0
+    }
+}
+
+//QUESTION / TODO?: It would be nice to wrap the environment_init all in a va_args call, unfortunately cbindgen
+// doesn't work with va_args and it's not worth requiring an additional build step for this.
+// Also, C doesn't have a natural way to express key-value pairs or named args, so we'd need to invent
+// one, which is probably more trouble than it's worth.  So I'll leave the stateful builder API the
+// way it is for now, but create a nicer API for Python using kwargs
+
+/// cbindgen:ignore
+static CURRENT_ENV_BUILDER: Mutex<(EnvInitState, Option<EnvBuilder>)> = std::sync::Mutex::new((EnvInitState::Uninitialized, None));
```

Comment:
May be implement builder in C in a straightforward way by passing instance of the builder as an argument to the setters? Is there any other reason to use global `CURRENT_ENV_BUILDER` except removing one of the arguments?

Reply:
To me it was a matter of API ergonomics & surface area.  Fundamentally configuring the platform environment is a stateful operation (ie. It's done only once at startup), so having stateful API functions doesn't actually sacrifice any flexibility.

Global data feels evil in an object-oriented language.  But C isn't an object oriented language, and sometimes making global state feel like an object is counter-productive.  This is just my opinion, however.

If you feel strongly, I can wrap this in something that feels more object-oriented.  I just wanted to keep the API surface area to a minimum and blend in with the conventions of the host language (C in this case).

Reply:
> To me it was a matter of API ergonomics & surface area. Fundamentally configuring the platform environment is a stateful operation (ie. It's done only once at startup), so having stateful API functions doesn't actually sacrifice any flexibility.

From my perspective it is a question of whether we need two different environments in the same process. For example if we use global instance of the `Environment` the unit tests which use different environment state cannot run in parallel. It is not fully the same but still relevant example is introducing reentrant versions of the C API functions into POSIX to overstep single thread implementation restrictions.

Reply:
That’s a very good point.  I didn’t consider the use case of multiple C / Python threads racing to init the Environment at the same time.  I figured the client app would perform the init before spawning child threads, but I see that’s not always possible.

I reworked the C API (and Python by extension, although it's hidden in Python) to use a builder object here: 586e2b2

As to whether we want to allow multiple environments to be active in the same process at the same time, that's a slightly different question.  Do you think that's important?

Reply:
In dc15fc3, I extensively reworked the environment API to permit custom environments to be initialized and used from any host language, and I provided conveniences to initialize a test environment that will not load / nor create any platform configuration files.

The singleton environment is now purely a convenience.